gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeBuilder.java

```diff
@@ -98,24 +98,33 @@ QueryNode buildExecutionTree(ExecutionContext executionContext, QueryNode filter
     // Try adding SortAndPaginateNode
     rootNode = checkAndAddSortAndPaginationNode(rootNode, executionContext);
 
-    // Fetch all other attributes, metric agg and time series data
-    if (!executionContext.getPendingSelectionSources().isEmpty()) {
+    if (isSingleSourceAndSame(executionContext.getPendingSelectionSources(), executionContext.getPendingMetricAggregationSources())) {
       rootNode =
           new SelectionNode.Builder(rootNode)
               .setAttrSelectionSources(executionContext.getPendingSelectionSources())
-              .build();
-      // Handle case where there is no order by but pagination still needs to be done
-      rootNode = checkAndAddSortAndPaginationNode(rootNode, executionContext);
-    }
-
-    if (!executionContext.getPendingMetricAggregationSources().isEmpty()) {
-      rootNode =
-          new SelectionNode.Builder(rootNode)
               .setAggMetricSelectionSources(executionContext.getPendingMetricAggregationSources())
               .build();
+      // Handle case where there is no order by but pagination still needs to be done
       rootNode = checkAndAddSortAndPaginationNode(rootNode, executionContext);
+    } else {
+      // Fetch all other attributes, metric agg and time series data
+      if (!executionContext.getPendingSelectionSources().isEmpty()) {
+        rootNode =
+            new SelectionNode.Builder(rootNode)
+                .setAttrSelectionSources(executionContext.getPendingSelectionSources())
+                .build();
+        // Handle case where there is no order by but pagination still needs to be done
+        rootNode = checkAndAddSortAndPaginationNode(rootNode, executionContext);
+      }
+
+      if (!executionContext.getPendingMetricAggregationSources().isEmpty()) {
```

Comment:
These ifs are not exclusive, right?

Reply:
I assume you're referring to these 2 ifs: 
```if (!executionContext.getPendingSelectionSources().isEmpty())```
and 
```if (!executionContext.getPendingMetricAggregationSources().isEmpty())``` 
They are not. We need to check for both selection and metric aggregations. Also, nothing changed in the Else section. It's the same as before this refactoring, I just moved it to else condition.